src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json
+{
+    public static partial class JsonSerializer
+    {
+        private static void HandleMetadataPropertyValue(ref Utf8JsonReader reader, ref ReadStack state)
+        {
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                ThrowHelper.ThrowJsonException_MetadataValueWasNotString();
+            }
+
+            MetadataPropertyName metadata = state.Current.MetadataProperty;
+            string key = reader.GetString()!;
+
+            if (metadata == MetadataPropertyName.Id)
+            {
+                state.ReferenceResolver.AddReference(key!, GetValueToPreserve(ref state));
+            }
+            else if (metadata == MetadataPropertyName.Ref)
+            {
+                state.Current.ReferenceId = key;
+            }
+
+            state.Current.ReadMetadataValue = false;
+        }
+
+        private static object GetValueToPreserve(ref ReadStack state)
+        {
+            return state.Current.IsProcessingProperty(ClassType.Dictionary) ?
+                state.Current.JsonPropertyInfo!.GetValueAsObject(state.Current.ReturnValue)! : state.Current.ReturnValue!;
+        }
+
+        internal static MetadataPropertyName GetMetadataPropertyName(ReadOnlySpan<byte> propertyName, ref ReadStack state, ref Utf8JsonReader reader)
+        {
+            if (propertyName[0] == '$')
+            {
+                switch (propertyName.Length)
+                {
+                    case 3:
+                        if (propertyName[1] == 'i' &&
+                            propertyName[2] == 'd')
+                        {
+                            return MetadataPropertyName.Id;
+                        }
+                        break;
+
+                    case 4:
+                        if (propertyName[1] == 'r' &&
+                            propertyName[2] == 'e' &&
+                            propertyName[3] == 'f')
+                        {
+                            return MetadataPropertyName.Ref;
+                        }
+                        break;
+
+                    case 7:
+                        // Only Preserved Arrays are allowed to read $values as metadata.
+                        if (state.Current.IsPreservedArray &&
+                            propertyName[1] == 'v' &&
```

Comment:
Would a `SequenceEqual` be faster here?

Reply:
I created a few benchmarks to compare using `SequenceEquals` against this, and seems that this approach is faster; I will share the benchmarks in a minute.

|                          Method | propertyName |     Mean |     Error |    StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|----------:|
|  GetMetadataUsingSequenceEquals |          $id | 6.438 ns | 0.0334 ns | 0.0312 ns | 6.434 ns | 6.367 ns | 6.496 ns |     - |     - |     - |         - |
| GetMetadataUsingSequenceEquals2 |          $id | 6.138 ns | 0.0712 ns | 0.0666 ns | 6.137 ns | 5.994 ns | 6.261 ns |     - |     - |     - |         - |
|         GetMetadataPropertyName |          $id | 2.406 ns | 0.0175 ns | 0.0155 ns | 2.403 ns | 2.386 ns | 2.444 ns |     - |     - |     - |         - |
|  GetMetadataUsingSequenceEquals |       $idddd | 6.879 ns | 0.0242 ns | 0.0214 ns | 6.884 ns | 6.846 ns | 6.912 ns |     - |     - |     - |         - |
| GetMetadataUsingSequenceEquals2 |       $idddd | 3.018 ns | 0.0103 ns | 0.0096 ns | 3.018 ns | 3.005 ns | 3.038 ns |     - |     - |     - |         - |
|         GetMetadataPropertyName |       $idddd | 2.017 ns | 0.0073 ns | 0.0061 ns | 2.016 ns | 2.010 ns | 2.031 ns |     - |     - |     - |         - |
|  GetMetadataUsingSequenceEquals |         $ref | 5.097 ns | 0.2196 ns | 0.2054 ns | 5.013 ns | 4.944 ns | 5.554 ns |     - |     - |     - |         - |
| GetMetadataUsingSequenceEquals2 |         $ref | 6.488 ns | 0.0668 ns | 0.0625 ns | 6.472 ns | 6.401 ns | 6.625 ns |     - |     - |     - |         - |
|         GetMetadataPropertyName |         $ref | 2.729 ns | 0.1098 ns | 0.3238 ns | 2.622 ns | 2.349 ns | 3.741 ns |     - |     - |     - |         - |
|  GetMetadataUsingSequenceEquals |      $values | 4.992 ns | 0.1134 ns | 0.0885 ns | 4.952 ns | 4.917 ns | 5.231 ns |     - |     - |     - |         - |
| GetMetadataUsingSequenceEquals2 |      $values | 7.152 ns | 0.1735 ns | 0.1622 ns | 7.130 ns | 6.829 ns | 7.543 ns |     - |     - |     - |         - |
|         GetMetadataPropertyName |      $values | 2.978 ns | 0.0322 ns | 0.0269 ns | 2.979 ns | 2.914 ns | 3.023 ns |     - |     - |     - |         - |
|  GetMetadataUsingSequenceEquals |       values | 8.093 ns | 0.0609 ns | 0.0570 ns | 8.101 ns | 7.979 ns | 8.181 ns |     - |     - |     - |         - |
| GetMetadataUsingSequenceEquals2 |       values | 2.834 ns | 0.0266 ns | 0.0249 ns | 2.840 ns | 2.778 ns | 2.884 ns |     - |     - |     - |         - |
|         GetMetadataPropertyName |       values | 1.943 ns | 0.0699 ns | 0.1866 ns | 1.839 ns | 1.775 ns | 2.494 ns |     - |     - |     - |         - |